server/src/main/java/org/opensearch/indices/replication/SegmentReplicator.java

```diff
@@ -197,12 +356,18 @@ int size() {
 
     void cancel(ShardId shardId, String reason) {
         onGoingReplications.cancelForShard(shardId, reason);
+        replicationCheckpointStats.remove(shardId);
+        latestPrimaryCheckpoint.remove(shardId);
     }
 
     SegmentReplicationTarget get(ShardId shardId) {
         return onGoingReplications.getOngoingReplicationTarget(shardId);
     }
 
+    ReplicationCheckpoint getLatestPrimaryCheckpoint(ShardId shardId) {
+        return latestPrimaryCheckpoint.get(shardId);
```

Comment:
pls use getOrDefault with ReplicationCheckpoint.Empty() to prevent NPEs, particularly in calls from the target service.